src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -220,13 +220,21 @@ private void CheckInvariants()
             solutionAttributes ??= _solutionAttributes;
             projectIds ??= ProjectIds;
             idToProjectStateMap ??= _projectIdToProjectStateMap;
-            options ??= Options.WithLanguages(GetRemoteSupportedProjectLanguages(idToProjectStateMap));
             analyzerReferences ??= AnalyzerReferences;
             projectIdToTrackerMap ??= _projectIdToTrackerMap;
             filePathToDocumentIdsMap ??= _filePathToDocumentIdsMap;
             dependencyGraph ??= _dependencyGraph;
             var newFrozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument.HasValue ? frozenSourceGeneratedDocument.Value : _frozenSourceGeneratedDocumentState;
 
+            // PERF: Only invoke WithLanguages if we have different set of project IDs (AddProject/RemoveProject operation)
```

Comment:
seems sane.  but i worry if it's possible to somehow have an Add+REmove processed at teh same time that woudl lead us to have the same count.  @jasonmalinowski for validation that we do things in tiny steps, so this sort of optimization is ok.

Reply:
I audited the existing callsites to validate that satisfy the first assertion. Definitely need @jasonmalinowski's confirmation on this one though.

Reply:
So for some reason we also have a ProjectIds collection we pass around, and that's immutable. I'm guessing you're safe to do the WithLanguages any time the ProjectId collection _object reference_ changes. The check is won't have false negatives where we fail to rerun WithLanguages; it could theoretically have a false positive where might run it if we manipulate the list and end up with the same original semantic value but I don't think we do that anywhere. (Or if we do, it's fine enough to not care since we'll still get quite the win.)

Reply:
Ah, that seems to be a much more better approach. Let me adjust.